components/tikv_alloc/src/trace.rs

```diff
@@ -0,0 +1,216 @@
+
+use std::num::NonZeroU64;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc, Weak,
+};
+use std::{
+    cmp,
+    fmt::{self, Display},
+};
+
+type HashMap<K, V> =
+    std::collections::HashMap<K, V, std::hash::BuildHasherDefault<fxhash::FxHasher>>;
+
+#[derive(Debug, PartialEq, Copy, Clone, Hash, Eq)]
+pub enum Id {
+    Name(&'static str),
+    Number(u64),
+}
+
+impl Id {
+    pub fn readable_name(&self) -> String {
+        match self {
+            Id::Name(s) => {
+                let mut s = s.to_string();
+                s.make_ascii_lowercase();
+                let mut res = String::with_capacity(s.len());
+                for p in s.split('_') {
+                    res.push_str(p);
+                    res.push(' ');
+                }
+                res
+            }
+            Id::Number(n) => n.to_string(),
+        }
+    }
+}
+
+impl From<&'static str> for Id {
+    #[inline]
+    fn from(s: &'static str) -> Self {
+        Id::Name(s)
+    }
+}
+
+impl From<u64> for Id {
+    #[inline]
+    fn from(id: u64) -> Self {
+        Id::Number(id)
+    }
+}
+
+impl From<NonZeroU64> for Id {
+    #[inline]
+    fn from(id: NonZeroU64) -> Self {
+        Id::Number(id.into())
+    }
+}
+
+impl Display for Id {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Id::Number(n) => write!(f, "{}", n),
+            Id::Name(n) => write!(f, "{}", n),
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+pub enum TraceEvent {
+    Add(usize),
+    Sub(usize),
+    Reset(usize),
+}
+
+pub trait MemoryTrace {
+    fn trace(&self, event: TraceEvent);
+    fn snapshot(&self, parent: &mut MemoryTraceSnapshot);
+    fn sub_trace(&self, id: Id) -> Arc<dyn MemoryTrace + Send + Sync>;
+    fn add_sub_trace(&mut self, id: Id, trace: Arc<dyn MemoryTrace + Send + Sync>);
+}
+
+pub struct MemoryTraceNode {
+    pub id: Id,
+    parent: Option<Weak<dyn MemoryTrace + Send + Sync>>,
+    trace: AtomicUsize,
+    children: HashMap<Id, Arc<dyn MemoryTrace + Send + Sync>>,
+}
+
+impl MemoryTraceNode {
+    pub fn new(id: impl Into<Id>) -> MemoryTraceNode {
+        MemoryTraceNode {
+            id: id.into(),
+            parent: None,
+            trace: std::sync::atomic::AtomicUsize::default(),
+            children: HashMap::default(),
+        }
+    }
+
+    fn parent_trace(&self, event: TraceEvent) {
+        if let Some(parent) = self.parent.as_ref() {
+            if let Some(p) = parent.upgrade() {
+                p.trace(event);
+            }
+        }
+    }
+
+    pub fn set_parent(&mut self, parent: &Arc<dyn MemoryTrace + Send + Sync>) {
+        self.parent = Some(Arc::downgrade(parent));
+    }
+}
+
+impl MemoryTrace for MemoryTraceNode {
+    fn trace(&self, event: TraceEvent) {
+        match event {
+            TraceEvent::Add(val) => {
+                self.trace.fetch_add(val, Ordering::Relaxed);
+                self.parent_trace(event);
```

Comment:
Tracing parent will make this operation more expansive. Every trace contains 3 atomic operations per node. Why can't it be updated during access?

Reply:
I think if we would use it for rate limiting, then it could be accessed very frequently. And surely we need to take care of the frequency of memory trace.

Reply:
/cc @Connor1996 How often can rate limiting be?

Tracing parent is not just expansive, it's also a waste. Not every updates need to be observed immediately. For example, if two queries allocate 1000 times and each allocates 10K. With tracing parent, the metrics will be updated 1000 times. But if update on read, then only read happen before every allocate will trigger calculation, which is as worse as tracing parent, otherwise, it performs better. If there is no read requests at all, then no calculation is necessary.

Reply:
I agree. We can sum it when necessary. So only atomics on leaf nodes needs to be updated.